src/join.jl

```diff
@@ -1,182 +1,15 @@
-            end
-            while j1 < rr && roweq(rkey, rperm[j1], rperm[j1+1])
-                j1 += 1
-            end
-            if typ !== :anti
-                if !grp
-                    for x=i:i1
-                        for y=j:j1
-                            push!(I, ks[lperm[x]])
-                            # optimized push! method for concat_tup
-                            _push!(Val{:both}(), f, data,
-                                   lout, rout, ldata, rdata,
-                                   lperm[x], rperm[y],
-                                   missing_instance(missingtype), missing_instance(missingtype))
-                        end
-                    end
-                else
-                    push!(I, ks[lperm[i]])
-                    group = init_group()
-                    for x=i:i1
-                        for y=j:j1
-                            group = accumulate(group, f(ldata[lperm[x]], rdata[rperm[y]]))
-                        end
-                    end
-                    push!(data, group)
-                    if keepkeys && i1+1 <= ll && roweq(lkey, lperm[i1], lperm[i1+1])
-                        # This means that the next key on the left is equal in the lkey sense
-                        # but different in the unique-key sense, so we start to make a new block again
-                        i = i1 + 1
-                        @goto nextgroup
-                    end
-                end
-            end
-            i = i1 + 1
-            j = j1 + 1
-        else
-            if typ === :outer
-                push!(I, rkey[rperm[j]])
-                if grp
-                    # empty group
-                    push!(data, init_group())
-                else
-                    _push!(Val{:right}(), f, data, lout, rout,
-                           ldata, rdata, 0, rperm[j], lnull, rnull)
-                    push!(lnull_idx, length(data))
-                end
-            end
-            j += 1
-        end
-    end
-
-    # finish up
-    if typ !== :inner
-        if (typ === :outer || typ === :left || typ === :anti) && i <= ll
-            append!(I, ks[lperm[i:ll]])
-            if grp
-                # empty group
-                append!(data, map(x->init_group(), i:ll))
-            else
-                append!(rnull_idx, (1:length(i:ll)) .+ length(data))
-                _append!(Val{:left}(), f, data, lout, rout,
-                       ldata, rdata, lperm[i:ll], 0, lnull, rnull)
-            end
-        elseif typ === :outer && j <= rr
-            append!(I, rkey[rperm[j:rr]])
-            if grp
-                # empty group
-                append!(data, map(x->init_group(), j:rr))
-            else
-                append!(lnull_idx, (1:length(j:rr)) .+ length(data))
-                _append!(Val{:right}(), f, data, lout, rout,
-                       ldata, rdata, 0, rperm[j:rr], lnull, rnull)
-            end
-        end
-    end
-    lnull_idx, rnull_idx
-end
```

Comment:
what have you done with my favorite secret function?

on a serious note this is awesome heroic work as always!

Reply:
I remember you mentioning it was the ugliest function you ever wrote :) I may need to bring some of that back to optimize the `f === concat_tup` case though.